### PR TITLE
IC-1728: Returned selectedCategoryIds from DraftReferrals

### PR DESCRIPTION
**Pull request by andrewburrowsmoj**
## What does this pull request do?

Adds serviceCategoryIds onto the Draft Referrals. 
This involves adding a connection to the referrals entity for selected service categories that the user has chosen.
For non-cohort referrals, the selected service category will always be connected to the draft referral (on creation).

## What is the intent behind these changes?

Provides functionality to allow user to specify chosen service categories for cohort referrals.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt [hunk 1]
```diff
@@ -141,14 +141,18 @@ class ReferralService(
     overrideCreatedAt: OffsetDateTime? = null,
     endOfServiceReport: EndOfServiceReport? = null,
   ): Referral {
+    val intervention = interventionRepository.getOne(interventionId)
+    val serviceCategories = intervention.dynamicFrameworkContract.contractType.serviceCategories
+    val selectedServiceCategories = if (serviceCategories.size == 1) serviceCategories.toMutableList() else mutableListOf()
```
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:146]: I thought it was best to assign the single service category on creating the referral if it's a non-cohort referral as it would otherwise require a separate call from the front-end to populate it, additional it's considered an invalid referral if it has no selected service category for non-cohort referral

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt [hunk 2]
```diff
@@ -67,6 +69,13 @@ class Referral(
 
   @OneToOne(mappedBy = "referral", optional = true) @Fetch(JOIN) var actionPlan: ActionPlan? = null,
 
+  @ManyToMany(fetch = FetchType.LAZY)
+  @JoinTable(
+    name = "referral_selected_service_category",
+    joinColumns = [JoinColumn(name = "referral_id")],
+    inverseJoinColumns = [JoinColumn(name = "service_category_id")]
+  )
+  var selectedServiceCategories: List<ServiceCategory>,
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:78]: we don't need to care about ordering. it's a presentation concern. we should use whatever is most efficient here.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt [hunk 3]
```diff
@@ -67,6 +69,13 @@ class Referral(
 
   @OneToOne(mappedBy = "referral", optional = true) @Fetch(JOIN) var actionPlan: ActionPlan? = null,
 
+  @ManyToMany(fetch = FetchType.LAZY)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:72]: how come this is manytomany instead of onetomany?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt (hunk 1 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:146]: yes absolutely. why is this cast to a `mutableList` vs a plain `list`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt (hunk 3 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:72]: A referral can have multiple service categories selected by the user, and a service category can be linked to multiple referrals.

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:72]: also, do we need to store an entity relationship here? i think we can just store the ids? might make the querying more efficient

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/authorization/SingleReferralEndpoints.kt [hunk 4]
```diff
@@ -51,7 +51,7 @@ class SingleReferralEndpoints : IntegrationTestBase() {
     return tokenFactory.createEncodedToken(userID = user.id, userName = user.userName, authSource = user.authSource)
   }
 
-  @ParameterizedTest
+  @ParameterizedTest(name = "{displayName} ({argumentsWithNames})")
```
Comment (tomsmyers) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/authorization/SingleReferralEndpoints.kt:54]: please can we split these changes into their own commit? 

the same goes for refactoring of the readme, and other logically seperate changes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt (hunk 1 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:146]: toList() does not include add() method which is required by hibernate otherwise it throws an `unsupportedOperationException`

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:146]: interesting - how come we dont' have to set the data type to MutableList on the entity class?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:78]: Ok I will change back to set here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt (hunk 1 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:146]: The datatype `List` is extended by `MutableList`. However the listOf() method returns a `EmptyList` object which doesn't contain `add()` methods

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralServiceTest.kt [hunk 5]
```diff
@@ -96,6 +97,23 @@ class ReferralServiceTest @Autowired constructor(
       SampleData.sampleReferral("X123456", "Harmony Living")
     )
     sampleIntervention = sampleReferral.intervention
+
+    sampleCohortIntervention = SampleData.persistIntervention(
```
Comment (rgforsyth) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralServiceTest.kt:101]: Not really an issue (I see rest of file is using sample data). But it may be useful when persisting data in tests going forward to use the factories if we are adding new things, so we can move away from sample data eventually.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt [hunk 6]
```diff
@@ -141,14 +141,18 @@ class ReferralService(
     overrideCreatedAt: OffsetDateTime? = null,
     endOfServiceReport: EndOfServiceReport? = null,
   ): Referral {
+    val intervention = interventionRepository.getOne(interventionId)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:144]: what's the reason to choose `getOne` over `findOne` in this instance?

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:144]: and what happens if the intervention id is invalid? perhaps we should throw ServerWebInputException?

``` 
val intervention = interventionRepository.findByIdOrNull(interventionId)
      ?: throw ServerWebInputException("invalid intervention id")
```

File: src/main/resources/db/local/V99_2_0__draft_referrals.sql [hunk 7]
```diff
@@ -7,6 +7,15 @@ values ('ac386c25-52c8-41fa-9213-fcf42e24b0b5', '98a42c61-c30f-4beb-8062-04033c3
...
        ('2a67075a-9c77-4103-9de0-63c4cfe3e8d6', '98a42c61-c30f-4beb-8062-04033c376e2d', '2021-01-12 14:46:21.987234+00', null, null, 'CRN22', '2021-04-01', '2500099998', 'Some information about the service user', 'Alex is currently sleeping on her aunt''s sofa', 'She uses a wheelchair', true, 'Spanish', true, 'She works Mondays 9am - midday', 'A danger to the elderly', true, 10, 'd0db50b0-4a50-4fc7-a006-9c97530e38b2', null);
 
+insert into referral_selected_service_category()
+values ('ac386c25-52c8-41fa-9213-fcf42e24b0b5', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('dfb64747-f658-40e0-a827-87b4b0bdcfed', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('745011c1-1ae5-45e4-99cb-6e1f2f8ccab9', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('d496e4a7-7cc1-44ea-ba67-c295084f1962', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('1219a064-709b-4b6c-a11e-10b8cb3966f6', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('037cc90b-beaa-4a32-9ab7-7f79136e1d27', '428ee70f-3001-4399-95a6-ad25eaaede16')
+        ('2a67075a-9c77-4103-9de0-63c4cfe3e8d6', '428ee70f-3001-4399-95a6-ad25eaaede16')
```
Comment (rgforsyth) [inline src/main/resources/db/local/V99_2_0__draft_referrals.sql:17]: missing semi-colon

Comment (rgforsyth) [reply src/main/resources/db/local/V99_2_0__draft_referrals.sql:17]: and commar's

File: src/main/resources/db/local/V99_2_1__sent_referrals.sql [hunk 8]
```diff
@@ -2,6 +2,11 @@ insert into referral (id, intervention_id, created_at, sent_at, sent_by_id, serv
 values ('81d754aa-d868-4347-9c0f-50690773014e', '98a42c61-c30f-4beb-8062-04033c376e2d', '2021-01-11 10:32:12.382884+00', '2021-01-14 15:56:45.382884+00', '2500099998', 'CRN23', '2021-04-01', '2500099998', 'Some information about the service user', 'Alex is currently sleeping on her aunt''s sofa', 'She uses a wheelchair', true, 'Spanish', true, 'She works Mondays 9am - midday', 'A danger to the elderly', true, 10, 'd0db50b0-4a50-4fc7-a006-9c97530e38b2', 'HD2123AC'),
        ('f89bd739-b9a2-482e-9947-12a793abcfb1', '3ccb511b-89b2-42f7-803b-304f54d85a24', '2021-01-11 10:32:12.382884+00', '2021-01-14 15:56:45.382884+00', '2500099998', 'CRN23', '2021-04-01', '2500099998', 'Some information about the service user', 'Alex is currently sleeping on her aunt''s sofa', 'She uses a wheelchair', true, 'Spanish', true, 'She works Mondays 9am - midday', 'A danger to the elderly', true, 10, 'd0db50b0-4a50-4fc7-a006-9c97530e38b2', 'AJ1827DR');
 
+insert into referral_selected_service_category()
+values ('81d754aa-d868-4347-9c0f-50690773014e', '428ee70f-3001-4399-95a6-ad25eaaede16')
```
Comment (rgforsyth) [inline src/main/resources/db/local/V99_2_1__sent_referrals.sql:6]: as above

File: src/main/resources/db/local/V99_2_2__assigned_referrals.sql [hunk 9]
```diff
@@ -3,6 +3,12 @@ values ('BBC2FEEE-2E86-4A33-8069-DD79A606B741', '08524319-7d5b-4b56-862a-bfe2c9a
        ('B59D3599-0681-466A-82B2-F6F957E46190', '90e35306-5d65-4901-bb36-c4c44c7d19f9', '2021-03-10 16:49:34.235464+00', '2021-03-10 16:50:34.235464+00', '11', 'CRN25', '2022-01-01', '11', '', '', '', false, null, false, null, '', true, 5, 'd0db50b0-4a50-4fc7-a006-9c97530e38b2', 'GH3927AC', '2021-03-10 17:50:34.235464+00', '6c4036b7-e87d-44fb-864f-5a06c1c492f3', '608955ae-52ed-44cc-884c-011597a77949'),
        ('a2a551aa-3d11-44b1-907b-42a028852bc1', 'f803445f-326c-4ef8-aee2-f7716d417832', '2021-04-22 09:00:00.000000+00', '2021-04-22 09:00:00.000000+00', '2500099998', 'CRN21', '2021-06-21', '2500099998', '', '', '', false, null, false, null, '', false, null, '64eb3c83-c562-4a63-9182-68a3a7725463', 'FH12376SO', '2021-04-22 09:00:00.000000+00', '6c4036b7-e87d-44fb-864f-5a06c1c492f3', '608955ae-52ed-44cc-884c-011597a77949');
 
+insert into referral_selected_service_category()
+values ('BBC2FEEE-2E86-4A33-8069-DD79A606B741', '428ee70f-3001-4399-95a6-ad25eaaede16')
+       ('B59D3599-0681-466A-82B2-F6F957E46190', '428ee70f-3001-4399-95a6-ad25eaaede16')
```
Comment (rgforsyth) [inline src/main/resources/db/local/V99_2_2__assigned_referrals.sql:8]: as above

File: src/main/resources/db/local/V99_2_0__draft_referrals.sql [hunk 10]
```diff
@@ -7,6 +7,15 @@ values ('ac386c25-52c8-41fa-9213-fcf42e24b0b5', '98a42c61-c30f-4beb-8062-04033c3
        ('037cc90b-beaa-4a32-9ab7-7f79136e1d27', '98a42c61-c30f-4beb-8062-04033c376e2d', '2021-01-12 14:46:21.987234+00', null, null, 'CRN21', null, '2500099998', null, null, null, null, null, null, null, null, null, null, null, null),
        ('2a67075a-9c77-4103-9de0-63c4cfe3e8d6', '98a42c61-c30f-4beb-8062-04033c376e2d', '2021-01-12 14:46:21.987234+00', null, null, 'CRN22', '2021-04-01', '2500099998', 'Some information about the service user', 'Alex is currently sleeping on her aunt''s sofa', 'She uses a wheelchair', true, 'Spanish', true, 'She works Mondays 9am - midday', 'A danger to the elderly', true, 10, 'd0db50b0-4a50-4fc7-a006-9c97530e38b2', null);
 
+insert into referral_selected_service_category()
```
Comment (rgforsyth) [inline src/main/resources/db/local/V99_2_0__draft_referrals.sql:10]: Do column names need to be specified if using brackets here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt (hunk 6 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:144]: getOne throws EntityNotFoundException which the ReferralController catches and rethrows it as the exception you described above.

I think that ServerWebInputException is outside the scope of this class?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/authorization/SingleReferralEndpoints.kt (hunk 4 above)
Comment (andrewburrowsmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/authorization/SingleReferralEndpoints.kt:54]: Sure, it's a habit of squashing everything. I'll be more vigilant in future!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt (hunk 3 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:72]: We decided on using entity relationship over UUID on slack

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt (hunk 6 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:144]: 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt [hunk 11]
```diff
@@ -49,7 +49,9 @@ data class DraftReferralDTO(
         serviceUser = ServiceUserDTO.from(referral.serviceUserCRN, referral.serviceUserData),
         serviceProvider = ServiceProviderDTO.from(contract.primeProvider),
         relevantSentenceId = referral.relevantSentenceId,
-        serviceCategoryId = contract.contractType.serviceCategories.elementAt(0).id,
+        // TODO: remove this once cohort referrals changes are complete
+        serviceCategoryId = if (referral.selectedServiceCategories?.isNotEmpty() == true) referral.selectedServiceCategories!!.elementAt(0).id else null,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DraftReferralDTO.kt:53]: not a blocker, i think this could be simplified to just `referral.selectedServiceCategories?.elementAt(0)`

File: src/main/resources/db/migration/V1_51__referral_selected_service_categories.sql [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+
+-- If there are any existing referrals then assume that the selected service categories are all of the ones present in the contract
+insert into referral_selected_service_category
+    select referral.id, service_category.id from referral, intervention, dynamic_framework_contract, contract_type, contract_type_service_category, service_category
+    where referral.intervention_id = intervention.id
+    and intervention.dynamic_framework_contract_id = dynamic_framework_contract.id
+    and dynamic_framework_contract.contract_type_id = contract_type.id
+    and contract_type.id = contract_type_service_category.contract_type_id
+    and contract_type_service_category.service_category_id = service_category.id
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/resources/db/migration/V1_51__referral_selected_service_categories.sql:18]: nice 👍 

missing trailing semi-colon. not sure if this has any impact.

**Review by tomsmyers — APPROVED** [review]
